### ISSUE TITLE
App Role authentication with `secret_id_bound_cidrs` fails to detect correct IP from X-Forwarded-For

### ISSUE THREAD

**Issue by oliverjanik** [hashicorp/vault#12334]
Vault version 1.8.1 installed from the official docker.

The vault is running in Fargate behind AWS ELB.

Server config:
```
listener "tcp" {  
  address = "0.0.0.0:8200"
  tls_disable = "true"
  x_forwarded_for_authorized_addrs = "1.2.3.4/32"
}
```

Approle config:
```
vault read auth/approle/role/role-name

Key                     Value                
bind_secret_id          true                 
local_secret_ids        false                
policies                ["kv-readonly"]      
secret_id_bound_cidrs   ["1.2.3.4/32"]
secret_id_num_uses      0                    
secret_id_ttl           0                    
token_bound_cidrs       []                   
token_explicit_max_ttl  0                    
token_max_ttl           900                  
token_no_default_policy false                
token_num_uses          0                    
token_period            0                    
token_policies          ["kv-readonly"]      
token_ttl               0                    
token_type              default              
```

When trying to authenticate with approle:
```
vault write auth/approle/login role_id=... secret_id=...

Error writing data to auth/approle/login: Error making API request.

URL: PUT https://vault.unifii.dev/v1/auth/approle/login
Code: 400. Errors:

* source address "10.0.1.230" unauthorized by CIDR restrictions on the role: %!w(<nil>)
```

Where the source address 10.0.1.230 is the internal ELB address. The client machine I'm connecting from has public IP 1.2.3.4 and is not hosted in AWS.

I've searched all issues around X-forwarded-for to no avail. Most closed unresolved.

looking at the code: https://github.com/hashicorp/vault/blob/861454e0ed1390d67ddaf1a53c1798e5e291728c/builtin/credential/approle/path_login.go#L271

it seems you're just grabbing connection remoteIP whereas I'd expect it would grab the true Client IP from X-Forwarded-For.

**Comment by Nineluj** [comment]
That seems like the intended behavior. [The docs](https://www.vaultproject.io/docs/configuration/listener/tcp#x_forwarded_for_authorized_addrs) say:
> x_forwarded_for_authorized_addrs - Specifies the list of source IP CIDRs for which an X-Forwarded-For header will be trusted

I assume that the purpose of this setting is to specify which LBs/proxies you trust to send a trustworthy X-Forwarded-For header. Have you tried using your LB's IP address `10.0.1.230` as the value for `x_forwarded_for_authorized_addrs`?

**Comment by oliverjanik** [comment]
You understand that I'm trying to white-list which external IPs are allowed to authenticate against an approle?

Are you saying that isn't supported? That seems like an oversight?

While-listing the LB does nothing for me.

**Comment by angelichorsey** [comment]
FWIW, `req.Connection.RemoteAddr` isn't from `net/http`. It's a struct in vault code and represents `logical.Request`. In this case, `RemoteAddr` is just a string that is set. Normally I think it's set after processing the base `http.Request` struct. In any case, it is something that always must be processed by vault. The snippet below shows how it is set to one of the addresses in the XFF header.
https://github.com/hashicorp/vault/blob/8e7fd9e794f1b3be560728c126a28fb7a84b6ed7/http/handler.go#L425-L455

Otherwise, I believe there is some confusion about the settings you have configured. In the listener config for vault, `x_forwarded_for_authorized_addrs` both enables processing of XFF, and it defines what addresses to trust for XFF. In this case you want to trust the address of the load balancer. This means that if clients could hit a vault node in a different manner, any provided XFF would be ignored since it isn't the load balancer CIDR.

The next bit of configuration where you specify `secret_id_bound_cidrs` is correct, as that defines which clients are authorized.

TL;DR: listener config needs to know what addresses to trust for XFF, and app-role config is the actual client CIDR that you want to grant acces to.

**Comment by oliverjanik** [comment]
I think I understand the explanation. however that still doesn't address my use case.

I already have an IP whitelist between ELB -> Vault. It's called AWS security group (aka firewall)

What I'd like to do only allow **specific** public IP to be able to authenticate with a **specific** app role. Any way to achieve that?

I thought that's what x-forwarded-for would do for me. Otherwise it seems like a not very useful setting..

**Comment by hsimon-hashicorp** [comment]
Hi folks, were there any other feedback or questions on this issue?

**Comment by oliverjanik** [comment]
I'll try what @angelichorsey has suggested. But the documentation needs to be way clearer how to achive this basic use-case.

I'll report in a few days.

**Comment by hghaf099** [comment]
Based on the above discussion, I am going to close this issue. Please feel free to reopen the issue or open a new one for further discussions.

**Comment by onetwopunch** [comment]
I'm facing the same issue with the Vault helm chart deploying to GKE with very basic configuration. I tried adding the LB IP to the `x_forwarded_for_authorized_cidrs` with no luck, but then noticed that the IP Vault is seeing actually comes from the node (host) IP, not the LB at all (they were only slightly different). But adding the node IPs to that list still did not make Vault see the client IP.
